src/drivers/fluid_opensles.c

```diff
@@ -0,0 +1,414 @@
+/* FluidSynth - A Software Synthesizer
+ *
+ * Copyright (C) 2003  Peter Hanappe and others.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public License
+ * as published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA
+ */
+
+/* fluid_opensles.c
+ *
+ * Audio driver for OpenSLES.
+ *
+ */
+
+#include "fluid_synth.h"
+#include "fluid_adriver.h"
+#include "fluid_settings.h"
+
+#if OPENSLES_SUPPORT
+
+#include <sys/time.h>
+#include <SLES/OpenSLES.h>
+#include <SLES/OpenSLES_Android.h>
+
+#define NUM_CHANNELS 2
+
+/** fluid_opensles_audio_driver_t
+ *
+ * This structure should not be accessed directly. Use audio port
+ * functions instead.
+ */
+typedef struct {
+  fluid_audio_driver_t driver;
+  SLObjectItf engine;
+  SLObjectItf output_mix_object;
+  SLObjectItf audio_player;
+  SLPlayItf audio_player_interface;
+  SLAndroidSimpleBufferQueueItf player_buffer_queue_interface;
+  
+  void* synth;
+  int period_frames;
+
+  int is_sample_format_float;
+  int use_callback_mode;
+
+  /* used only by callback mode */
+  short* sles_buffer_short;
+  float* sles_buffer_float;
+  fluid_audio_func_t callback;
+  /* used only by non-callback mode */
+  fluid_thread_t *thread;
+
+  int cont;
+  long next_expected_enqueue_time;
+  
+  double sample_rate;
+} fluid_opensles_audio_driver_t;
+
+
+fluid_audio_driver_t* new_fluid_opensles_audio_driver(fluid_settings_t* settings,
+						   fluid_synth_t* synth);
+fluid_audio_driver_t* new_fluid_opensles_audio_driver2(fluid_settings_t* settings,
+						    fluid_audio_func_t func, void* data);
+void delete_fluid_opensles_audio_driver(fluid_audio_driver_t* p);
+void fluid_opensles_audio_driver_settings(fluid_settings_t* settings);
+static fluid_thread_return_t fluid_opensles_audio_run(void* d);
+static void opensles_callback(SLAndroidSimpleBufferQueueItf caller, void *pContext);
+void adjust_latency(fluid_opensles_audio_driver_t* dev);
+void process_fluid_buffer(fluid_opensles_audio_driver_t* dev);
+
+void fluid_opensles_audio_driver_settings(fluid_settings_t* settings)
+{
+  fluid_settings_register_int(settings, "audio.opensles.use-callback-mode", 1, 0, 1,
```

Comment:
What is your motivation to provide a "non-callback" mode, given the workaround you are attempting in the `adjust_latency` function?

Reply:
I reviewed the entire things and figured out that I don't really need this one, so removed the entire "non-callback" related code. That resulted in lots of removal of code. (more comment at 5034a57)